rewrite-java/src/main/java/org/openrewrite/java/JavaSourceVisitor.java

```diff
@@ -45,32 +46,60 @@
+     * <P><P>
+     * A terminating element is one of the following:
+     * <P><P>
+     * <li>A static class declaration</li>
+     * <li>An enumeration declaration</li>
+     * <li>An interface declaration</li>
+     * <li>An annotation declaration</li>
+     *
+     * @param base The point within the tree that is used to establish the "base scope".
+     * @param child A point within the tree that will be traversed (up the tree) looking for an intersection with the base scope.
+     * @return If the base scope is found within the context of the child path, this method will return true, otherwise false
+     */
+    default boolean isInSameNameScope(Cursor base, Cursor child) {
+        //First establish the base scope by finding the first enclosing element.
+        Tree baseScope = base.getPathAsStream()
                 .filter(t -> t instanceof J.Block ||
                         t instanceof J.MethodDecl ||
                         t instanceof J.Try ||
                         t instanceof J.ForLoop ||
-                        t instanceof J.ForEachLoop).findAny()
-                .map(higherNameScope -> lower.getPathAsStream()
-                        .filter(t -> {
-                            takeWhile.set(takeWhile.get() && (
-                                    !(t instanceof J.ClassDecl) ||
-                                            (((J.ClassDecl) t).getKind() instanceof J.ClassDecl.Kind.Class &&
-                                                    !((J.ClassDecl) t).hasModifier("static"))));
-                            return takeWhile.get();
-                        })
-                        .anyMatch(higherNameScope::equals))
-                .orElse(false);
+                        t instanceof J.ForEachLoop)
+                .findFirst().orElseThrow(() -> new IllegalArgumentException("The base cursor does not have an scoped context."));
+
+        //Now walk up the child path looking for the base scope.
+        for (Iterator<Tree> it = child.getPath(); it.hasNext(); ) {
+            Tree childScope = it.next();
+            if (childScope instanceof J.ClassDecl) {
+                J.ClassDecl childClass = (J.ClassDecl) childScope;
+                if (!(childClass.getKind() instanceof J.ClassDecl.Kind.Class) || childClass.hasModifier("static")) {
+                    //Short circuit the search if a terminating element is encountered.
+                    break;
+                }
+            }
+            if (baseScope.isScope(childScope)) {
+                return true;
+            }
+        }
+
+        return false;
     }
 
     /**
-     * @param lower The cursor of the lower scoped tree element to check.
+     * Returns true if the passed child cursor is within the same name scope as the current cursor.
+     *
+     * @param child The cursor of the child tree element to check.
      * @return Whether this cursor shares the same name scope as {@code scope}.
```

Comment:
This comment is difficult to understand. 

You don't need both the @return bit and the  "Returns true if..."  description. Particularly since the existing @return description is so confusing and mentions a symbol, "scope", not present anywhere in the signature.  Do one or the other. 

We should use the term "lexical scope" rather than "name scope", I believe the former will generally be better understood. 

I suggest something like this:
```
/*
 * @param child a cursor to an AST element
 * @return true if the child is present within the lexical scope of the AST element currently being visited
 */
```
But that description isn't perfect. Does "within the lexical scope of the AST element currently being visited" mean that this function would return true for peer/sibling elements within the same scope? Or only if the current AST element's scope _encloses_ child? See if you can find a way to make that clear.

Reply:
Agreed. I will try to clean this up.